docarray/doc_index/backends/elastic_doc_index.py

```diff
@@ -0,0 +1,271 @@
+    ###############################################
+
+    def python_type_to_db_type(self, python_type: Type) -> Any:
+        """Map python type to database type."""
+        for allowed_type in ELASTIC_PY_VEC_TYPES:
+            if issubclass(python_type, allowed_type):
+                return np.ndarray
+
+        if python_type == docarray.typing.ID:
+            return None  # TODO(johannes): handle this
+
+        raise ValueError(f'Unsupported column type for {type(self)}: {python_type}')
+
+    def _index(self, column_data_dic, **kwargs):
+        # not needed, we implement `index` directly
+        ...
+
+    def num_docs(self) -> int:
+        return self._client.count(index=self._db_config.index_name)['count']
+
+    def _del_items(self, doc_ids: Sequence[str]):
+        ...
+
+    def _get_items(self, doc_ids: Sequence[str]) -> Sequence[TSchema]:
+        ...
+
+    def execute_query(self, query: List[Tuple[str, Dict]], *args, **kwargs) -> Any:
+        ...
+
+    def _find_batched(
+        self,
+        query: np.ndarray,
+        search_field: str,
+        limit: int,
+        **kwargs,
+    ) -> FindResultBatched:
+        ...
+
+    @composable
+    def _find(
+        self, query: np.ndarray, search_field: str, limit: int, **kwargs
+    ) -> FindResult:
+        ...
+
+    @composable
+    def _filter(
+        self,
+        *args,
+        **kwargs,
+    ) -> DocumentArray:
+        ...
+
+    def _filter_batched(
+        self,
+        filter_queries: Any,
+        limit: int,
+        **kwargs,
+    ) -> List[DocumentArray]:
+        ...
+
+    def _text_search(
+        self,
+        query: str,
+        search_field: str,
+        limit: int,
+        **kwargs,
+    ) -> FindResult:
+        ...
+
+    def _text_search_batched(
+        self,
+        queries: Sequence[str],
+        search_field: str,
+        limit: int,
+        **kwargs,
+    ) -> FindResultBatched:
+        ...
+
+    ####################################################
+    # Optional overrides                               #
+    ####################################################
+
+    def index(self, docs: Union[BaseDocument, Sequence[BaseDocument]], **kwargs):
+        """Index a document into the store"""
+        if kwargs:
+            raise ValueError(f'{list(kwargs.keys())} are not valid keyword arguments')
+        doc_seq = docs if isinstance(docs, Sequence) else [docs]
+        requests = []
+
+        for doc in doc_seq:
+            request = {
+                '_index': self._db_config.index_name,
+                '_id': doc.id,
+                'blob': doc.to_base64(),
+            }
+            for col_name, col in self._columns.items():
+                if not col.config:
+                    continue
+                request[col_name] = doc[col_name].tolist()
+            requests.append(request)
+
+        self._send_requests(request)
+        self._refresh(self._db_config.index_name)
```

Comment:
Refresh is used to update internal lucene index so that the new documents are available for search. Let's make this as runtime config. Depending on the number of docs the refresh operation can be expensive and unnecessary. If you're creating an index for the first then doing a refresh after completing all indexing operations is more efficient.